enterprise_access/apps/content_assignments/api.py

```diff
@@ -272,49 +277,44 @@ def allocate_assignments(assignment_configuration, learner_emails, content_key,
         assignment_configuration,
         learner_emails_to_allocate,
         content_key,
+        lms_user_ids_by_email,
     )
-
-    # Keep a running list of all existing assignments that will need to be included in bulk update.
-    #
-    # Note: Since the duplicates are just references to the same assignment object, and django model instances are
-    # hashable (on PK), they can be de-duplicated by blinndly adding them to this set.
-    existing_assignments_needs_update = set()
-
     # Maintain a set of emails with existing records - we know we don't have to create
     # new assignments for these.
     learner_emails_with_existing_assignments = set()
 
-    # Try to populate lms_user_id field on any existing assignments found.  We already ran this function on these
-    # assignments (when they were created in a prior request), but time has passed since then so the outcome might be
-    # different this time. It's technically possible some learners have registered since the last request.
-    existing_assignments_needs_update.update(_try_populate_assignments_lms_user_id(existing_assignments))
+    # Keep a running list of all existing assignments that will need to be included in bulk update.
+    existing_assignments_needs_update = set()
 
-    # Time has passed since the last time this assignment was allocated/re-allocated. Therefore, it's entirely
-    # possible a new run has been published. We assume the intent of the admin re-allocating this content is
-    # that they want to assign the NEW run. This step to update preferred_course_run_key is required in order
+    # This step to find and update the preferred_course_run_key is required in order
     # for nudge emails to target the start date of the new run.
     preferred_course_run_key = _get_preferred_course_run_key(assignment_configuration, content_key)
 
     # Split up the existing assignment records by state
     for assignment in existing_assignments:
-        learner_emails_with_existing_assignments.add(assignment.learner_email.lower())
+        if not assignment.lms_user_id:
+            existing_lms_user_id = lms_user_ids_by_email.get(assignment.learner_email.lower())
+            if existing_lms_user_id:
+                assignment.lms_user_id = existing_lms_user_id
+                existing_assignments_needs_update.add(assignment)
+
+        if assignment.state == LearnerContentAssignmentStateChoices.EXPIRED and assignment.lms_user_id is not None:
+            # If the existing assignment is expired and has an lms_user_id, it has a retired/expired email address
+            # that we want to change based on our lookup of lms_user_id -> email.
+            assignment_email_from_lms_user_id = emails_by_lms_user_id.get(assignment.lms_user_id)
+            if assignment_email_from_lms_user_id is not None:
+                assignment.learner_email = assignment_email_from_lms_user_id
+                existing_assignments_needs_update.add(assignment)
```

Comment:
This is the crux of the change - we now identify existing assignments via either type of user identifier (email or lms_user_id).